packages/vm/src/evm/evm.ts

```diff
@@ -579,12 +729,168 @@ export default class EVM {
+    }
+
+    const oldRefund = this._refund
+
+    await this._state.checkpoint()
+    this._transientStorage.checkpoint()
+    if (this.DEBUG) {
+      debug('-'.repeat(100))
+      debug(`message checkpoint`)
+    }
+
+    let result
+    if (this.DEBUG) {
+      const { caller, gasLimit, to, value, delegatecall } = message
+      debug(
+        `New message caller=${caller} gasLimit=${gasLimit} to=${
+          to?.toString() ?? 'none'
+        } value=${value} delegatecall=${delegatecall ? 'yes' : 'no'}`
+      )
+    }
+    if (message.to) {
+      if (this.DEBUG) {
+        debug(`Message CALL execution (to: ${message.to})`)
+      }
+      result = await this._executeCall(message as MessageWithTo)
+    } else {
+      if (this.DEBUG) {
+        debug(`Message CREATE execution (to undefined)`)
+      }
+      result = await this._executeCreate(message)
+    }
+    if (this.DEBUG) {
+      const { gasUsed, exceptionError, returnValue } = result.execResult
+      debug(
+        `Received message execResult: [ gasUsed=${gasUsed} exceptionError=${
+          exceptionError ? `'${exceptionError.error}'` : 'none'
+        } returnValue=0x${short(returnValue)} gasRefund=${result.gasRefund ?? 0} ]`
+      )
+    }
+    const err = result.execResult.exceptionError
+    // This clause captures any error which happened during execution
+    // If that is the case, then set the _refund tracker to the old refund value
+    if (err) {
+      this._refund = oldRefund
+      result.execResult.selfdestruct = {}
+    }
+    result.gasRefund = this._refund
+    if (err) {
+      if (this._common.gteHardfork(Hardfork.Homestead) || err.error != ERROR.CODESTORE_OUT_OF_GAS) {
+        result.execResult.logs = []
+        await this._state.revert()
+        this._transientStorage.revert()
+        if (this.DEBUG) {
+          debug(`message checkpoint reverted`)
+        }
+      } else {
+        // we are in chainstart and the error was the code deposit error
+        // we do like nothing happened.
+        await this._state.commit()
+        this._transientStorage.commit()
+        if (this.DEBUG) {
+          debug(`message checkpoint committed`)
+        }
+      }
+    } else {
+      await this._state.commit()
+      this._transientStorage.commit()
+      if (this.DEBUG) {
+        debug(`message checkpoint committed`)
+      }
+    }
+    await this._emit('afterMessage', result)
+
+    return result
+  }
+
+  /**
+   * Bound to the global VM and therefore
+   * shouldn't be used directly from the evm class
+   * @hidden
+   */
+  async runCode(opts: RunCodeOpts): Promise<ExecResult> {
+    const block = opts.block ?? Block.fromBlockData({}, { common: this._common })
+    this._block = block
+
+    // Backwards compatibility
+    const txContext: TxContext = {
+      gasPrice: opts.gasPrice ?? BigInt(0),
+      origin: opts.origin ?? opts.caller ?? Address.zero(),
+    }
+    this._tx = txContext
```

Comment:
instead of defining the const, just set it directly?

```suggestion
    this._tx = {
      gasPrice: opts.gasPrice ?? BigInt(0),
      origin: opts.origin ?? opts.caller ?? Address.zero(),
    }
```

Reply:
good point, fixed